packages/service-proxy/src/mixins/service.mixin.ts

```diff
@@ -0,0 +1,198 @@
+ * called <a href="#ServiceMixinDoc">ServiceMixinDoc</a>
+ *
+ */
+// tslint:disable-next-line:no-any
+export function ServiceMixin<T extends Class<any>>(superClass: T) {
+  return class extends superClass {
+    // A mixin class has to take in a type any[] argument!
+    // tslint:disable-next-line:no-any
+    constructor(...args: any[]) {
+      super(...args);
+    }
+
+    /**
+     * Add a service to this application.
+     *
+     * @param provider The service provider to register.
+     *
+     * ```ts
+     * export interface GeocoderService {
+     *   geocode(address: string): Promise<GeoPoint[]>;
+     * }
+     *
+     * export class GeocoderServiceProvider implements Provider<GeocoderService> {
+     *   constructor(
+     *     @inject('datasources.geocoder')
+     *     protected datasource: juggler.DataSource = new GeocoderDataSource(),
+     *   ) {}
+     *
+     *   value(): Promise<GeocoderService> {
+     *     return getService(this.datasource);
+     *   }
+     * }
+     *
+     * app.serviceProvider(GeocoderServiceProvider);
+     * ```
+     */
+    serviceProvider<S>(provider: Class<Provider<S>>): void {
+      const serviceName = provider.name.replace(/Provider$/, '');
+      const repoKey = `services.${serviceName}`;
+      this.bind(repoKey)
+        .toProvider(provider)
+        .tag('service');
+    }
+
+    /**
+     * Add a component to this application. Also mounts
+     * all the components services.
+     *
+     * @param component The component to add.
+     *
+     * ```ts
+     *
+     * export class ProductComponent {
+     *   controllers = [ProductController];
+     *   repositories = [ProductRepo, UserRepo];
+     *   providers = {
+     *     [AUTHENTICATION_STRATEGY]: AuthStrategy,
+     *     [AUTHORIZATION_ROLE]: Role,
+     *   };
+     * };
+     *
+     * app.component(ProductComponent);
+     * ```
+     */
+    public component(component: Class<{}>) {
+      super.component(component);
+      this.mountComponentRepository(component);
+    }
+
+    /**
+     * Get an instance of a component and mount all it's
+     * services. This function is intended to be used internally
+     * by component()
+     *
+     * @param component The component to mount services of
+     */
+    mountComponentRepository(component: Class<{}>) {
+      const componentKey = `components.${component.name}`;
+      const compInstance = this.getSync(componentKey);
+
+      if (compInstance.serviceProviders) {
+        for (const provider of compInstance.serviceProviders) {
+          this.serviceProvider(provider);
+        }
+      }
+    }
+  };
+}
+
+/**
+ * Interface for an Application mixed in with ServiceMixin
+ */
+export interface AppWithServices extends Application {
+  // tslint:disable-next-line:no-any
+  serviceProvider<S>(provider: Class<Provider<S>>): void;
+  component(component: Class<{}>): void;
+  mountComponentRepository(component: Class<{}>): void;
```

Comment:
It should be `mountComponentServiceProviders` too.

Reply:
Actually, I prefer to call this method `mountComponentServices`. In the future, if we add support for `app.service(serviceCtor: Class)`, then I'd like `mountComponentServices` to handle both service classes and providers.